src/AggregateFunctions/ReservoirSamplerDeterministic.h

```diff
@@ -162,6 +164,17 @@ class ReservoirSamplerDeterministic
             DB::writePODBinary(samples[i], buf);
     }
 
+    void sortIfNeeded()
+    {
+        if (sorted)
+            return;
+        sorted = true;
```

Comment:
Do we still need this flag now, when we have explicit finalization? The above exceptions could become `assert(std::is_sorted(...))`.

This is another sign of the problem that I'm talking about -- finalization is not explicit enough, we're not really sure when the state is finalized and when it's not, so in some places we add additional checks, and in some places we don't... So when reading this code in the future, it's going to be even harder to answer this question.

Reply:
Let's replace it to assert. I agree it will be more reasonable.